xdsl/transforms/csl_stencil_bufferize.py

```diff
@@ -379,6 +381,92 @@ def match_and_rewrite(self, op: arith.Constant, rewriter: PatternRewriter, /):
         )
 
 
+@dataclass(frozen=True)
+class InjectApplyOutsIntoLinalgOuts(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.ApplyOp, rewriter: PatternRewriter, /):
+        yld = op.done_exchange.block.last_op
+        assert isinstance(yld, csl_stencil.YieldOp)
+        new_dest: list[SSAValue] = []
+        new_yield_args: list[SSAValue] = []
+        additional_args: list[SSAValue] = []
+        to_remove: list[Operation] = []
+
+        for arg, yld_arg in zip(op.dest, yld.arguments, strict=True):
+            if (
+                not isinstance(yld_arg, OpResult)
+                or not isinstance(yld_arg.op, bufferization.ToMemrefOp)
+                or not isinstance(yld_arg.op.tensor, OpResult)
+                or not isinstance(
+                    linalg_op := yld_arg.op.tensor.op,
+                    linalg.NamedOpBase | linalg.Generic,
+                )
+                or not isa(arg_t := arg.type, AnyMemRefType)
+                or not isa(yld_arg.type, AnyMemRefType)
+            ):
+                new_dest.append(arg)
+                new_yield_args.append(yld_arg)
+                continue
+            additional_args.append(arg)
+            if len(yld_arg.uses) == 1:
+                to_remove.append(yld_arg.op)
+
+            arg = op.done_exchange.block.insert_arg(
+                arg.type, len(op.done_exchange.block.args)
+            )
+            arg_to_tensor = to_tensor_op(arg, writable=True)
+            symmetric_offsets = tuple(
+                (s - d) // 2
+                for s, d in zip(
+                    arg_t.get_shape(), yld_arg.type.get_shape(), strict=True
+                )
+            )
```

Comment:
Could you please put a note here? I'm not 100% sure I understand why we do `(s - d) // 2`

Reply:
This calculates the offset of the core data, assuming a symmetric (i.e. even) number of ghost cells in each direction. Done some renaming and added a comment as suggested.